### PR TITLE
[HttpKernel] Fix collecting uploaded files

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no     
| Deprecations? | no
| Tests pass?   | yes    
| Fixed tickets | #29178
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php [hunk 1]
```diff
@@ -59,6 +59,14 @@ public function collect(Request $request, Response $response, \Exception $except
 
         $requestFiles = array();
         foreach ($request->files->all() as $files) {
+            if (null === $files) {
+                continue;
```
[the remark this answers is not in the document]
Comment (xabbuh) [reply src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: Can this happen?

[the remark this answers is not in the document]
Comment (ro0NL) [reply src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: yes, that was the error i actually got (compared to the error from the ticket)

[the remark this answers is not in the document]
Comment (ro0NL) [reply src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: for

```php
$form = $f->create(FileType::class, null, ['required' => false, 'multiple' => false]);
return $this->render('index.html.twig', ['form' => $form->createView()]);
```

```html
{{ form_start(form, {attr: {novalidate: true}}) }}
    {{ form_widget(form) }}
    <input type="submit">
{{ form_end(form) }}
```

**Review by xabbuh — DISMISSED** [review]

File: src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php [hunk 2]
```diff
@@ -59,7 +59,18 @@ public function collect(Request $request, Response $response, \Exception $except
...
+            }
+
+            if (!\is_array($files)) {
+                $files = array($files);
+            }
+
             foreach ($files as $fileName => $fileData) {
+                if (null === $fileData) {
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: fixes `Call to a member function getClientOriginalName() on null`

for

```php
$form = $f->createBuilder()
    ->add('post', FileType::class, ['required' => false, 'mapped' => false])
    ->getForm();
return $this->render('index.html.twig', ['form' => $form->createView()]);
```

File: src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php [hunk 3]
```diff
@@ -59,12 +59,30 @@ public function collect(Request $request, Response $response, \Exception $except
...
-                    'name' => $fileData->getClientOriginalName(),
-                    'mimetype' => $fileData->getMimeType(),
-                    'size' => $fileData->getSize(),
-                );
+                if (null === $fileData) {
+                    continue;
+                }
+
+                if (!is_array($fileData)) {
+                    $fileData = array($fileData);
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: fixes `Call to a member function getClientOriginalName() on array` for nested multiple form.

**Review by gharlan — APPROVED** [review]
Thank you @ro0NL!

**Comment by ro0NL** [comment]
Yes, all good now :+1:

File: src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php (hunk 3 above)
Comment (ro0NL) [reply src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: ouch.. I think you're right :D

**Comment by ro0NL** [comment]
Per https://github.com/symfony/symfony/pull/29183#discussion_r232655483 

logic should be rercursified.

status: needs work

File: src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php (hunk 3 above)
Comment (ro0NL) [reply src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: fixed

**Comment by ro0NL** [comment]
status: needs review

**Review by javiereguiluz — APPROVED** [review]
Tested successfully in a 4.2-beta1 app. Thanks again Roland!

**Review by xabbuh — APPROVED** [review]

**Comment by stof** [comment]
is this a bug affecting only 4.2 ?

**Comment by ro0NL** [comment]
@stof yes

**Review by stof — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @ro0NL.

**Pull request merged**